text/0000-osstring-string-interface.md

```diff
@@ -0,0 +1,239 @@
+- Feature Name: osstring_string_interface
+- Start Date: 2015-10-05
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a string-like API to the `OsString` and `OsStr` types.  This RFC
+focuses on creating a string-like interface, as opposed to RFC #1307,
+which focuses more on container-like features.
+
+# Motivation
+
+As mentioned in the `std::ffi::os_str` documentation: "**Note**: At
+the moment, these types are extremely bare-bones, usable only for
+conversion to/from various other string types. Eventually these types
+will offer a full-fledged string API."  This is intended as a step in
+that direction.
+
+Having an ergonomic way to manipulate OS strings is needed to allow
+programs to easily handle non-UTF-8 data received from the operating
+system.  Currently, it is common for programs to just convert OS data
+to `String`s, which leads to undesirable panics in the unusual case
+where the input is not UTF-8.  For example, currently, calling rustc
+with a non-UTF-8 command line argument will result in an immediate
+panic.  Fixing that in a way that actually handles non-UTF-8 data
+correctly (as opposed to, for example, just interpreting it lossily as
+UTF-8) would be very difficult with the current OS string API.  Most
+of the functions proposed here were motivated by the OS string
+processing needs of rustc.
+
+# Detailed design
+
+## `OsString`
+
+`OsString` will get the following new method:
+```rust
+/// Converts an `OsString` into a `String`, avoiding a copy if possible.
+///
+/// Any non-Unicode sequences are replaced with U+FFFD REPLACEMENT CHARACTER.
+pub fn into_string_lossy(self) -> String;
+
+```
+
+This is analogous to the existing `OsStr::to_string_lossy` method, but
+transfers ownership.  This operation can be done without a copy if the
+`OsString` contains UTF-8 data or if the platform is Windows.
+
+## `OsStr`
+
+OsStr will get the following new methods:
+```rust
+/// Returns true if the string starts with a valid UTF-8 sequence
+/// equal to the given `&str`.
+fn starts_with_str(&self, prefix: &str) -> bool;
```

Comment:
When adding string-like APIs to `OsStr`, it may be best to try to stick to original `str` APIs as much as possible, while also allowing all kinds of fun functionality for `OsStr`. Along those lines, perhaps this API could look like:

```rust
fn starts_with<S: AsRef<OsStr>>(&self, prefix: S) -> bool;
```

That should cover this use case as well as something like `os_str.starts_with(&other_os_string)`.